sound/soc/sof/ops.c

```diff
@@ -161,3 +161,92 @@ void snd_sof_dsp_panic(struct snd_sof_dev *sdev, u32 offset)
 	snd_sof_trace_notify_for_error(sdev);
 }
 EXPORT_SYMBOL(snd_sof_dsp_panic);
+
+/* This is for getting ref count for a DSP core and power on it if needed */
+int snd_sof_dsp_core_get(struct snd_sof_dev *sdev, u32 core_idx)
+{
+	u32 target_cores_mask = sdev->enabled_cores_mask | BIT(core_idx);
+	int ret = 0;
+
+	mutex_lock(&sdev->cores_status_mutex);
+
+	/* return if already powered on */
+	if (sdev->core_refs[core_idx] > 0) {
+		dev_vdbg(sdev->dev, "core_get: enabled_cores_mask 0x%x, core_refs[%d] %d, no need to power up\n",
+			 sdev->enabled_cores_mask, core_idx, sdev->core_refs[core_idx]);
+		sdev->core_refs[core_idx]++;
+		goto done;
```

Comment:
the logic would be so much simpler if you took the mutex inside the power_up()/power_down()